zrml/neo-swaps/src/types/liquidity_tree.rs

```diff
@@ -0,0 +1,1205 @@
+        if node.is_leaf() {
+            self.mutate_node(index, |node| {
+                node.fees = node.fees.checked_add_res(&node.lazy_fees)?;
+                Ok(())
+            })?;
+        } else {
+            // The lazy fees that will be propagated down the tree.
+            let mut remaining_lazy_fees =
+                node.descendant_stake.bmul_bdiv(node.lazy_fees, node.total_stake()?)?;
+            // The fees that stay at this node.
+            let fees = node.lazy_fees.checked_sub_res(&remaining_lazy_fees)?;
+            self.mutate_node(index, |node| {
+                node.fees = node.fees.checked_add_res(&fees)?;
+                Ok(())
+            })?;
+            let child_indices = self.children(index)?;
+            // Unwrap below can't fail since we're unwrapping from an array of length 2.
+            match child_indices.get(0).unwrap_or(&None) {
+                Some(lhs_index) => {
+                    self.mutate_node(*lhs_index, |lhs_node| {
+                        let child_lazy_fees = lhs_node
+                            .total_stake()?
+                            .bmul_bdiv(remaining_lazy_fees, descendant_stake)?;
+                        lhs_node.lazy_fees =
+                            lhs_node.lazy_fees.checked_add_res(&child_lazy_fees)?;
+                        remaining_lazy_fees =
+                            remaining_lazy_fees.checked_sub_res(&child_lazy_fees)?;
+                        Ok(())
+                    })?;
+                }
+                None => (),
+            }
+            // Unwrap below can't fail since we're unwrapping from an array of length 2.
+            match child_indices.get(1).unwrap_or(&None) {
+                Some(rhs_index) => {
+                    self.mutate_node(*rhs_index, |rhs_node| {
+                        rhs_node.lazy_fees =
+                            rhs_node.lazy_fees.checked_add_res(&remaining_lazy_fees)?;
+                        Ok(())
+                    })?;
+                }
+                None => (),
+            }
+        }
+        self.mutate_node(index, |node| {
+            node.lazy_fees = Zero::zero();
+            Ok(())
+        })?;
+        Ok(())
+    }
+
+    fn children(&self, index: u32) -> Result<[Option<u32>; 2], DispatchError> {
+        let calculate_child =
+            |child_index: u32| Some(child_index).filter(|&i| i < self.node_count());
+        let left_child_index = index.checked_mul_res(&2)?.checked_add_res(&1)?;
+        let left_child = calculate_child(left_child_index);
+        let right_child_index = left_child_index.checked_add_res(&1)?;
+        let right_child = calculate_child(right_child_index);
+        Ok([left_child, right_child])
+    }
+
+    fn parent_index(&self, index: u32) -> Option<u32> {
+        if index == 0 {
+            None
+        } else {
+            // Won't ever fail, always returns `Some(...)`.
+            index.checked_sub(1)?.checked_div(2)
+        }
+    }
```

Comment:
I really don't think the fact that `i - 1` doesn't underflow if `i != 0` requires a `log` and a `debug_assert`. Honestly, I'm not sure why I didn't just do `index.checked_sub(1)?.checked_div` instead of the whole if-else.

Reply:
Maybe you wanted to remind the reader for the distinct case for the root node. 

Me as a reviewer I just saw `Won't ever fail` and instantly thought to protect this case programmatically by a debug_assert.

Reply:
otherwise you might not notice that your code "broke".

Reply:
Fixed it by removing the comment. b46d145e0d7dc2f417049e55a2176bc13ecf4669

Reply:
I really don't think there's any point in verifying an assumption that simple.